File: packages/adapter-claude-code/src/events.ts
import type { DomainEvent, Item, Usage } from '@harness/contracts'
import { JsonRpcValueSchema } from '@harness/proc'
import { z } from 'zod'

/**
 * The Claude Code stream-json envelope, as the binary actually emits it.
 *
 * Typed by hand from observed output rather than from a published schema —
 * unlike Codex, there is no generator. That makes this the fragile end of the
 * adapter, so every field is optional and anything unrecognised becomes an
 * `unknown` item instead of throwing. Verified against claude-code 2.1.220.
 */

const JsonObjectSchema = z.record(z.string(), JsonRpcValueSchema)
const ResultPartSchema = z.union([z.string(), z.object({ text: z.string().optional() })])
const ToolResultContentSchema = z.union([z.string(), z.array(ResultPartSchema)])
const ContentBlockSchema = z.object({
  type: z.string(),
  text: z.string().optional(),
  thinking: z.string().optional(),
  id: z.string().optional(),
  name: z.string().optional(),
  input: JsonObjectSchema.optional(),
  tool_use_id: z.string().optional(),
  content: ToolResultContentSchema.optional(),
  is_error: z.boolean().optional(),
})
const ClaudeUsageSchema = z.object({
  input_tokens: z.number().optional(),
  output_tokens: z.number().optional(),
  cache_read_input_tokens: z.number().optional(),
  cache_creation_input_tokens: z.number().optional(),
})
export const ClaudeEventSchema = z.object({
  type: z.string().optional(),
  subtype: z.string().optional(),
  session_id: z.string().optional(),
  uuid: z.string().optional(),
  message: z
    .object({
      id: z.string().optional(),
      role: z.string().optional(),
      model: z.string().optional(),
      content: z.array(ContentBlockSchema).optional(),
      usage: ClaudeUsageSchema.optional(),
    })
    .optional(),
  result: z.string().optional(),
  is_error: z.boolean().optional(),
  duration_ms: z.number().optional(),
  total_cost_usd: z.number().optional(),
  usage: ClaudeUsageSchema.optional(),
})

export type ClaudeEvent = z.infer<typeof ClaudeEventSchema>
type ContentBlock = z.infer<typeof ContentBlockSchema>
type ToolResultContent = z.infer<typeof ToolResultContentSchema>
type ClaudeUsage = z.infer<typeof ClaudeUsageSchema>

/** Tools whose call is really a shell command, so it reads as one. */
const SHELL_TOOLS = new Set(['Bash', 'PowerShell'])
/** Tools that write to disk. */
const EDIT_TOOLS = new Set(['Edit', 'MultiEdit', 'Write', 'NotebookEdit'])

/**
 * Translate one envelope into domain events.
 *
 * Returns a list because a single assistant message can carry text, thinking
 * and several tool calls at once — Claude Code batches where Codex streams.
 *
 * `tools` remembers each call by its tool_use id so the result that follows
 * lands on the same item: the call turns from started into completed (or
 * failed) with its output attached, the way the saved-history replay already
 * reads. Without the map a result has nothing to attach to and stands alone.
 */
export function toDomainEvents(
  event: ClaudeEvent,
  turnId: string,
  tools?: Map<string, Item>,
): DomainEvent[] {
  const at = Date.now()

  if (event.type === 'assistant' && event.message?.content) {
    return event.message.content.flatMap((block, index) => {
      const id = assistantBlockId(event, block, index, at)
      const item = blockToItem(block, id, turnId, at)
      if (!item) return []
      if (block.type === 'tool_use' && block.id && tools) {
        tools.set(block.id, item)
        return [{ type: 'item.completed' as const, item: { ...item, status: 'started' } }]
      }
      return [{ type: 'item.completed' as const, item }]
    })
  }

  if (event.type === 'user' && event.message?.content) {
    // Tool results come back as user messages. They belong to the tool call
    // that produced them, not to a new exchange, so they are not rendered as
    // the user speaking.
    return event.message.content.flatMap((block) => {
      if (block.type !== 'tool_result') return []
      const text = flattenContent(block.content)
      const call = block.tool_use_id ? tools?.get(block.tool_use_id) : undefined
      if (call && block.tool_use_id) {
        tools?.delete(block.tool_use_id)
        return [
          { type: 'item.completed' as const, item: completedCall(call, text, block.is_error) },
        ]
      }
      if (!text) return []
      return [
        {
          type: 'item.completed' as const,
          item: {
            id: `${block.tool_use_id ?? at}-result`,
            turnId,
            type: block.is_error ? ('error' as const) : ('tool_call' as const),
            status: 'completed' as const,
            text,
            createdAt: at,
          },
        },
      ]
    })
  }

  if (event.type === 'result') {
    const events: DomainEvent[] = []
    const usage = toUsage(event.usage, event.total_cost_usd)
    if (usage) events.push({ type: 'usage.updated', usage })
    events.push({
      type: 'turn.completed',
      turnId,
      status: event.is_error ? 'failed' : 'completed',
    })
    return events
  }

  return []
}

function assistantBlockId(
  event: ClaudeEvent,
  block: ContentBlock,
  index: number,
  fallback: number,
): string {
  if (block.type === 'tool_use') {
    if (block.id) return `${block.id}-call`
  }
  return `${event.message?.id ?? event.uuid ?? fallback}-${block.type}-${index}`
}

function blockToItem(
  block: ContentBlock,
  id: string,
  turnId: string,
  at: number,
): Item | undefined {
  const base = { id, turnId, status: 'completed' as const, createdAt: at }

  switch (block.type) {
    case 'text': {
      return block.text
        ? { ...base, type: 'message', role: 'assistant', text: block.text }
        : undefined
    }

    case 'thinking': {
      return block.thinking ? { ...base, type: 'reasoning', text: block.thinking } : undefined
    }

    case 'tool_use':
      return { ...base, ...toolUseItemFields(block.name ?? 'tool', block.input ?? {}) }

    default:
      return undefined
  }
}

type ToolInput = Record<string, unknown>

/**
 * What a tool_use block contributes to its item. Shell tools read as a
 * command, edit tools as a file change carrying the edit as a unified diff,
 * and everything else as a tool call whose text names the tool on the first
 * line and carries its input as one JSON line — the transcript's convention
 * for every provider.
 */
export function toolUseItemFields(
  name: string,
  input: ToolInput,
):
  | { type: 'command'; command: string }
  | { type: 'file_change'; path: string; text?: string; linesAdded?: number; linesRemoved?: number }
  | { type: 'tool_call'; text: string } {
  if (SHELL_TOOLS.has(name)) {
    return { type: 'command', command: String(input['command'] ?? name) }
  }
  if (EDIT_TOOLS.has(name)) {
    const path = String(input['file_path'] ?? input['notebook_path'] ?? '')
    const diff = editDiff(name, path, input)
    return {
      type: 'file_change',
      path,
      ...(diff
        ? {
            text: diff,
            linesAdded: diff.split('\n').filter((line) => /^\+(?!\+\+ )/.test(line)).length,
            linesRemoved: diff.split('\n').filter((line) => /^-(?!-- )/.test(line)).length,
          }
        : {}),
    }
  }
  const argumentLine = Object.keys(input).length ? JSON.stringify(input) : undefined
  return { type: 'tool_call', text: argumentLine ? `${name}\n${argumentLine}` : name }
}

/** A tool call whose result arrived: same item, finished, output attached. */
function completedCall(call: Item, output: string, failed: boolean | undefined): Item {
  const status = failed ? 'failed' : 'completed'
  if (call.type === 'command') return { ...call, status, ...(output ? { text: output } : {}) }
  if (call.type === 'file_change') {
    // The edit itself is the diff already on the item; a success message adds
    // nothing, but an error explains why the file did not change.
    return {
      ...call,
      status,
      ...(failed && output ? { text: [call.text, output].filter(Boolean).join('\n\n') } : {}),
    }
  }
  return { ...call, status, ...(output ? { text: `${call.text ?? ''}\n${output}` } : {}) }
}

/** The edit tools describe a change as strings; the transcript wants a diff. */
function editDiff(name: string, path: string, input: ToolInput): string | undefined {
  const lines = (value: unknown) =>
    typeof value === 'string' ? value.replace(/\n$/, '').split('\n') : []
  const hunks: Array<{ removed: string[]; added: string[] }> = []
  if (name === 'Write') {
    hunks.push({ removed: [], added: lines(input['content']) })
  } else if (name === 'Edit') {
    hunks.push({ removed: lines(input['old_string']), added: lines(input['new_string']) })
  } else if (name === 'MultiEdit' && Array.isArray(input['edits'])) {
    for (const edit of input['edits'] as unknown[]) {
      const change = edit as Record<string, unknown>
      hunks.push({ removed: lines(change['old_string']), added: lines(change['new_string']) })
    }
  } else if (name === 'NotebookEdit') {
    hunks.push({
      removed: [],
      added: input['edit_mode'] === 'delete' ? [] : lines(input['new_source']),
    })
  }
  const body = hunks
    .filter((hunk) => hunk.removed.length > 0 || hunk.added.length > 0)
    .flatMap(({ removed, added }) => [
      `@@ -${removed.length ? `1,${removed.length}` : '0,0'} +${added.length ? `1,${added.length}` : '0,0'} @@`,
      ...removed.map((line) => `-${line}`),
      ...added.map((line) => `+${line}`),
    ])
  if (body.length === 0) return undefined
  return [
    `diff --git a/${path} b/${path}`,
    name === 'Write' ? '--- /dev/null' : `--- a/${path}`,
    `+++ b/${path}`,
    ...body,
  ].join('\n')
}

/** Tool results are sometimes a string, sometimes a content-block array. */
function flattenContent(content: ToolResultContent | undefined): string {
  if (content === undefined) return ''
  if (typeof content === 'string') return content
  return content
    .map((part) => (typeof part === 'string' ? part : (part.text ?? '')))
    .join('')
    .trim()
}

export function toUsage(usage: ClaudeUsage | undefined, costUsd?: number): Usage | undefined {
  if (!usage) return undefined
  const input = usage.input_tokens ?? 0
  const output = usage.output_tokens ?? 0
  const cached = usage.cache_read_input_tokens ?? 0
  return {
    inputTokens: input,
    cachedInputTokens: cached,
    outputTokens: output,
    // Claude Code does not report reasoning tokens separately.
    reasoningTokens: 0,
    totalTokens: input + output + cached + (usage.cache_creation_input_tokens ?? 0),
    inputIncludesCached: false,
    ...(!(costUsd === undefined) ? { costUsd } : {}),
  }
}

File: crates/harness-adapter-claude-code/src/events.rs
use harness_protocol::{DomainEvent, Item, ItemStatus, ItemType, MessageRole, TurnStatus, Usage};
use serde_json::Value;

const SHELL_TOOLS: &[&str] = &["Bash", "PowerShell"];
const EDIT_TOOLS: &[&str] = &["Edit", "Write", "NotebookEdit"];

/// Map one Claude Code stream-json envelope captured from 2.1.220.
/// Every lookup is tolerant because this CLI surface has no published schema.
pub fn map_domain_events(event: &Value, turn_id: &str, created_at: f64) -> Vec<DomainEvent> {
    match string(event, "type") {
        Some("assistant") => assistant_events(event, turn_id, created_at),
        Some("user") => tool_result_events(event, turn_id, created_at),
        Some("result") => result_events(event, turn_id),
        _ => Vec::new(),
    }
}

pub fn map_usage(usage: Option<&Value>, cost_usd: Option<f64>) -> Option<Usage> {
    let usage = usage?.as_object()?;
    let input_tokens = number(usage.get("input_tokens"));
    let output_tokens = number(usage.get("output_tokens"));
    let cached_input_tokens = number(usage.get("cache_read_input_tokens"));
    let cache_creation_tokens = number(usage.get("cache_creation_input_tokens"));
    Some(Usage {
        input_tokens,
        cached_input_tokens,
        output_tokens,
        reasoning_tokens: 0.0,
        total_tokens: input_tokens + output_tokens + cached_input_tokens + cache_creation_tokens,
        cost_usd,
        context_window: None,
    })
}

fn assistant_events(event: &Value, turn_id: &str, created_at: f64) -> Vec<DomainEvent> {
    let Some(content) = event
        .get("message")
        .and_then(|message| message.get("content"))
        .and_then(Value::as_array)
    else {
        return Vec::new();
    };
    let id_root = event
        .get("message")
        .and_then(|message| string(message, "id"))
        .or_else(|| string(event, "uuid"))
        .map(str::to_owned)
        .unwrap_or_else(|| format!("{created_at:.0}"));
    content
        .iter()
        .enumerate()
        .filter_map(|(index, block)| {
            map_block(block, format!("{id_root}-{index}"), turn_id, created_at)
        })
        .map(|item| DomainEvent::ItemCompleted { item })
        .collect()
}

fn tool_result_events(event: &Value, turn_id: &str, created_at: f64) -> Vec<DomainEvent> {
    let Some(content) = event
        .get("message")
        .and_then(|message| message.get("content"))
        .and_then(Value::as_array)
    else {
        return Vec::new();
    };
    content
        .iter()
        .filter(|block| string(block, "type") == Some("tool_result"))
        .filter_map(|block| {
            let text = flatten_content(block.get("content"));
            (!text.is_empty()).then(|| DomainEvent::ItemCompleted {
                item: Item {
                    id: format!(
                        "{}-result",
                        string(block, "tool_use_id")
                            .map(str::to_owned)
                            .unwrap_or_else(|| format!("{created_at:.0}"))
                    ),
                    turn_id: turn_id.into(),
                    item_type: if boolean(block, "is_error") {
                        ItemType::Error
                    } else {
                        ItemType::ToolCall
                    },
                    status: ItemStatus::Completed,
                    role: None,
                    text: Some(text),
                    command: None,
                    exit_code: None,
                    duration_ms: None,
                    path: None,
                    lines_added: None,
                    lines_removed: None,
                    created_at,
                },
            })
        })
        .collect()
}

fn result_events(event: &Value, turn_id: &str) -> Vec<DomainEvent> {
    let mut events = Vec::new();
    if let Some(usage) = map_usage(
        event.get("usage"),
        event.get("total_cost_usd").and_then(Value::as_f64),
    ) {
        events.push(DomainEvent::UsageUpdated { usage });
    }
    events.push(DomainEvent::TurnCompleted {
        turn_id: turn_id.into(),
        status: if boolean(event, "is_error") {
            TurnStatus::Failed
        } else {
            TurnStatus::Completed
        },
    });
    events
}

fn map_block(block: &Value, id: String, turn_id: &str, created_at: f64) -> Option<Item> {
    let item = |item_type, role, text, command, path| Item {
        id,
        turn_id: turn_id.into(),
        item_type,
        status: ItemStatus::Completed,
        role,
        text,
        command,
        exit_code: None,
        duration_ms: None,
        path,
        lines_added: None,
        lines_removed: None,
        created_at,
    };
    match string(block, "type")? {
        "text" => nonempty(block, "text").map(|text| {
            item(
                ItemType::Message,
                Some(MessageRole::Assistant),
                Some(text.into()),
                None,
                None,
            )
        }),
        "thinking" => nonempty(block, "thinking")
            .map(|text| item(ItemType::Reasoning, None, Some(text.into()), None, None)),
        "tool_use" => {
            let name = string(block, "name").unwrap_or("tool");
            let input = block.get("input");
            if SHELL_TOOLS.contains(&name) {
                Some(item(
                    ItemType::Command,
                    None,
                    None,
                    Some(
                        input
                            .and_then(|input| string(input, "command"))
                            .unwrap_or(name)
                            .into(),
                    ),
                    None,
                ))
            } else if EDIT_TOOLS.contains(&name) {
                Some(item(
                    ItemType::FileChange,
                    None,
                    None,
                    None,
                    Some(
                        input
                            .and_then(|input| string(input, "file_path"))
                            .unwrap_or_default()
                            .into(),
                    ),
                ))
            } else {
                Some(item(
                    ItemType::ToolCall,
                    None,
                    Some(name.into()),
                    None,
                    None,
                ))
            }
        }
        _ => None,
    }
}

fn flatten_content(content: Option<&Value>) -> String {
    match content {
        Some(Value::String(text)) => text.clone(),
        Some(Value::Array(parts)) => parts
            .iter()
            .filter_map(|part| match part {
                Value::String(text) => Some(text.as_str()),
                Value::Object(_) => string(part, "text"),
                _ => None,
            })
            .collect::<String>()
            .trim()
            .into(),
        _ => String::new(),
    }
}

fn string<'a>(value: &'a Value, key: &str) -> Option<&'a str> {
    value.get(key).and_then(Value::as_str)
}

fn nonempty<'a>(value: &'a Value, key: &str) -> Option<&'a str> {
    string(value, key).filter(|text| !text.is_empty())
}

fn boolean(value: &Value, key: &str) -> bool {
    value.get(key).and_then(Value::as_bool).unwrap_or(false)
}

fn number(value: Option<&Value>) -> f64 {
    value.and_then(Value::as_f64).unwrap_or(0.0)
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    const AT: f64 = 1_000.0;

    #[test]
    fn maps_captured_assistant_blocks_without_vendor_fields() {
        let events = map_domain_events(
            &json!({
                "type": "assistant",
                "message": {
                    "id": "msg_2",
                    "role": "assistant",
                    "model": "claude-fable-5",
                    "content": [
                        { "type": "thinking", "thinking": "Checking." },
                        { "type": "text", "text": "Found it." },
                        { "type": "tool_use", "id": "tu1", "name": "Bash", "input": { "command": "node -v" } },
                        { "type": "tool_use", "id": "tu2", "name": "Edit", "input": { "file_path": "src/app.ts" } },
                        { "type": "tool_use", "id": "tu3", "name": "WebSearch", "input": { "query": "docs" } }
                    ]
                }
            }),
            "turn-1",
            AT,
        );

        assert!(matches!(
            &events[0],
            DomainEvent::ItemCompleted { item }
                if item.item_type == ItemType::Reasoning && item.text.as_deref() == Some("Checking.")
        ));
        assert!(matches!(
            &events[1],
            DomainEvent::ItemCompleted { item }
                if item.item_type == ItemType::Message
                    && item.role == Some(MessageRole::Assistant)
                    && item.text.as_deref() == Some("Found it.")
        ));
        assert!(matches!(
            &events[2],
            DomainEvent::ItemCompleted { item }
                if item.item_type == ItemType::Command
                    && item.command.as_deref() == Some("node -v")
        ));
        assert!(matches!(
            &events[3],
            DomainEvent::ItemCompleted { item }
                if item.item_type == ItemType::FileChange
                    && item.path.as_deref() == Some("src/app.ts")
        ));
        assert!(matches!(
            &events[4],
            DomainEvent::ItemCompleted { item }
                if item.item_type == ItemType::ToolCall
                    && item.text.as_deref() == Some("WebSearch")
        ));
    }

    #[test]
    fn maps_captured_tool_results_and_flattens_content_blocks() {
        let events = map_domain_events(
            &json!({
                "type": "user",
                "message": {
                    "content": [
                        {
                            "type": "tool_result",
                            "tool_use_id": "tu1",
                            "content": [{ "type": "text", "text": "not " }, "found"],
                            "is_error": true
                        },
                        { "type": "text", "text": "not rendered as a user message" }
                    ]
                }
            }),
            "turn-1",
            AT,
        );

        assert_eq!(events.len(), 1);
        assert!(matches!(
            &events[0],
            DomainEvent::ItemCompleted { item }
                if item.id == "tu1-result"
                    && item.item_type == ItemType::Error
                    && item.text.as_deref() == Some("not found")
        ));
    }

    #[test]
    fn maps_captured_result_usage_cost_and_failure() {
        let events = map_domain_events(
            &json!({
                "type": "result",
                "is_error": true,
                "duration_ms": 400,
                "total_cost_usd": 0.04,
                "usage": {
                    "input_tokens": 2,
                    "output_tokens": 4,
                    "cache_read_input_tokens": 24787,
                    "cache_creation_input_tokens": 9297
                }
            }),
            "turn-1",
            AT,
        );

        assert!(matches!(
            &events[0],
            DomainEvent::UsageUpdated { usage }
                if usage.cached_input_tokens == 24787.0
                    && usage.total_tokens == 34090.0
                    && usage.cost_usd == Some(0.04)
        ));
        assert_eq!(
            events[1],
            DomainEvent::TurnCompleted {
                turn_id: "turn-1".into(),
                status: TurnStatus::Failed,
            }
        );
    }

    #[test]
    fn ignores_unknown_and_partially_malformed_envelopes() {
        assert!(map_domain_events(&json!({ "type": "future" }), "turn-1", AT).is_empty());
        assert!(map_domain_events(&json!({}), "turn-1", AT).is_empty());
        assert!(
            map_domain_events(
                &json!({ "type": "assistant", "message": { "content": "wrong" } }),
                "turn-1",
                AT,
            )
            .is_empty()
        );
    }
}
